File: lib/src/main/java/com/qwwuyu/lib/gson/DateTypeAdapter.java
package com.qwwuyu.lib.gson;

import com.google.gson.TypeAdapter;
import com.google.gson.internal.bind.util.ISO8601Utils;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonToken;
import com.google.gson.stream.JsonWriter;

import java.io.IOException;
import java.text.DateFormat;
import java.text.ParsePosition;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

class DateTypeAdapter extends TypeAdapter<Date> {
    private final List<DateFormat> dateFormats = new ArrayList<>();

    public DateTypeAdapter() {
        dateFormats.add(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault()));
        dateFormats.add(DateFormat.getDateTimeInstance(DateFormat.DEFAULT, DateFormat.DEFAULT));
        if (!Locale.getDefault().equals(Locale.US)) {
            dateFormats.add(DateFormat.getDateTimeInstance(DateFormat.DEFAULT, DateFormat.DEFAULT, Locale.US));
        }
    }

    @Override
    public Date read(JsonReader in) throws IOException {
        JsonToken peek = in.peek();
        if (peek == JsonToken.NULL) {
            in.nextNull();
            return null;
        } else if (peek == JsonToken.NUMBER) {
            return new Date(in.nextLong());
        }
        return deserializeToDate(in.nextString());
    }

    private synchronized Date deserializeToDate(String json) {
        if (json == null || json.length() == 0) {
            return null;
        }
        try {
            if (json.matches("^[\\d]{13}$")) {
                return new Date(Long.parseLong(json));
            }
        } catch (Exception ignored) {
        }
        for (DateFormat dateFormat : dateFormats) {
            try {
                return dateFormat.parse(json);
            } catch (Exception ignored) {
            }
        }
        try {
            return ISO8601Utils.parse(json, new ParsePosition(0));
        } catch (Exception e) {
            //throw new JsonSyntaxException(json, e);
        }
        return null;
    }

    @Override
    public synchronized void write(JsonWriter out, Date value) throws IOException {
        if (value == null) {
            out.nullValue();
            return;
        }
        String dateFormatAsString = dateFormats.get(0).format(value);
        out.value(dateFormatAsString);
    }
}

File: lib/src/main/kotlin/com/qwwuyu/lib/gson/DateTypeAdapter.kt
package com.qwwuyu.lib.gson

import com.google.gson.TypeAdapter
import com.google.gson.internal.bind.util.ISO8601Utils
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonToken
import com.google.gson.stream.JsonWriter
import java.io.IOException
import java.text.DateFormat
import java.text.ParsePosition
import java.text.SimpleDateFormat
import java.util.*

internal class DateTypeAdapter : TypeAdapter<Date?>() {
    private val dateFormats: MutableList<DateFormat> = ArrayList()

    init {
        dateFormats.add(SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault()))
        dateFormats.add(DateFormat.getDateTimeInstance(DateFormat.DEFAULT, DateFormat.DEFAULT))
        if (Locale.getDefault() != Locale.US) {
            dateFormats.add(DateFormat.getDateTimeInstance(DateFormat.DEFAULT, DateFormat.DEFAULT, Locale.US))
        }
    }

    @Throws(IOException::class)
    override fun read(`in`: JsonReader): Date? {
        val peek = `in`.peek()
        if (peek == JsonToken.NULL) {
            `in`.nextNull()
            return null
        } else if (peek == JsonToken.NUMBER) {
            return Date(`in`.nextLong())
        }
        return deserializeToDate(`in`.nextString())
    }

    @Synchronized
    private fun deserializeToDate(json: String?): Date? {
        if (json == null || json.isEmpty()) {
            return null
        }
        try {
            if (json.matches("^[\\d]{13}$".toRegex())) {
                return Date(json.toLong())
            }
        } catch (ignored: Exception) {
        }
        for (dateFormat in dateFormats) {
            try {
                return dateFormat.parse(json)
            } catch (ignored: Exception) {
            }
        }
        try {
            return ISO8601Utils.parse(json, ParsePosition(0))
        } catch (e: Exception) {
            //throw new JsonSyntaxException(json, e);
        }
        return null
    }

    @Synchronized
    @Throws(IOException::class)
    override fun write(out: JsonWriter, value: Date?) {
        if (value == null) {
            out.nullValue()
            return
        }
        val dateFormatAsString = dateFormats[0].format(value)
        out.value(dateFormatAsString)
    }
}